### ISSUE TITLE
Filter CodeLists by any attribute

### ISSUE THREAD

**Issue by danielhuppmann** [IAMconsortium/nomenclature#233]
Building on #226, we should now extend the `filter()` method and make it usable for any **CodeList** instance (not just **RegionCodeList**).

The use case that I have in mind is filtering a list of variables to have only those that are *required*, e.g. that have an attribute like
```yaml
- Final Energy|Electricity:
    description: Final energy consumption of electricity (including on-site solar
      PV), excluding transmission/distribution losses
    unit: EJ/yr
    required: true
```

See a real-world use case [here](https://github.com/iiasa/engage-gains-mapping/blob/ccb1cfd237a021ac0df1f37dae67a8fef5a4df42/definitions/variable/variable.yaml#L39).

**Comment by phackstock** [comment]
I guess the interface for this would look something like this then:

```python

class CodeList:

    def filter(self, attribute, value) -> CodeList:
        matched = [code.attribute == value for code in self]
        # return a new CodeList(matched)
```

The above example wouldn't work exactly like this but is this what you had in mind @danielhuppmann?

**Comment by danielhuppmann** [comment]
Well, I'd rather follow the pyam `filter()` signature

```python
class CodeList:

    def filter(self, **kwargs) -> CodeList:
        keep = np.ones(len(self), dtype=bool)
        for attribute, value in kwargs:
            keep = np.logical_and(keep, [code.attribute == value for code in self])
            # return a new CodeList(keep)
```

This would also be compatible with the existing signature (if a user wrote `filter(hierarchy="common")`, not `filter("common")`).

**Comment by phackstock** [comment]
Ok, good point, that is cleaner and more powerful.

**Comment by phackstock** [comment]
Closing as this was addressed by #236
